typing/typedecl.ml

```diff
@@ -1351,6 +1361,85 @@ let transl_exception env sext =
   let newenv = Env.add_extension ~check:true ext.ext_id ext.ext_type env in
     ext, newenv
 
+type native_repr_attribute =
+  | Native_repr_attr_absent
+  | Native_repr_attr_present of native_repr_kind
+  | Native_repr_attr_present_with_type
+    of native_repr_kind * Parsetree.core_type
+
+let get_native_repr_attribute core_type =
+  match
+    List.filter
+      (fun (n, _) ->
+         match n.Location.txt with
+         | "unboxed" | "untagged" -> true
+         | _ -> false)
+      core_type.ptyp_attributes
+  with
+  | [] ->
+    Native_repr_attr_absent
+  | _ :: (n, _) :: _ ->
+    raise (Error (n.Location.loc, Multiple_native_repr_attributes))
+  | [(n, payload)] ->
+    let kind = if n.txt = "unboxed" then Unboxed else Untagged in
+    match payload with
+    | PStr [] ->
+      Native_repr_attr_present kind
+    | PTyp core_type ->
+      Native_repr_attr_present_with_type (kind, core_type)
+    | _ ->
+      raise (Error (n.Location.loc,
+                    Invalid_native_repr_attribute_payload kind))
+
+let native_repr_of_type env kind ty =
+  match kind, (Ctype.expand_head_opt env ty).desc with
+  | Untagged, Tconstr (path, _, _) when Path.same path Predef.path_int ->
+    Some Untagged_int
+  | Unboxed, Tconstr (path, _, _) when Path.same path Predef.path_float ->
+    Some Unboxed_float
+  | Unboxed, Tconstr (path, _, _) when Path.same path Predef.path_int32 ->
+    Some (Unboxed_integer Pint32)
+  | Unboxed, Tconstr (path, _, _) when Path.same path Predef.path_int64 ->
+    Some (Unboxed_integer Pint64)
+  | Unboxed, Tconstr (path, _, _) when Path.same path Predef.path_nativeint ->
+    Some (Unboxed_integer Pnativeint)
+  | _ ->
+    None
+
+let make_native_repr env core_type ty =
+  match get_native_repr_attribute core_type with
+  | Native_repr_attr_absent -> Same_as_ocaml_repr
+  | Native_repr_attr_present kind ->
+    begin match native_repr_of_type env kind ty with
+    | None ->
+      raise (Error (core_type.ptyp_loc, Cannot_unbox_or_untag_type kind))
+    | Some repr -> repr
+    end
+  | Native_repr_attr_present_with_type (kind, core_type') ->
+    let ty' = (Typetexp.transl_simple_type env false core_type').ctyp_type in
+    match
+      native_repr_of_type env kind ty,
+      native_repr_of_type env kind ty'
+    with
+    | _, None ->
+      raise (Error (core_type'.ptyp_loc, Cannot_unbox_or_untag_type kind))
+    | None, Some repr -> repr
```

Comment:
> Is this check done elsewhere, or do we intend to trust the programmer annotation?

We have to trust the programmer annotation. But I can't remember why I wanted this and this a dubious feature indeed so I'll remove it